File: linked_list/150_intersection_of_two_ll.cpp
// 160. Intersection of Two Linked Lists

#include <unordered_set>

#include "LinkedList.h"
using namespace std;

class Solution {
   public:
    ListNode* getIntersectionNode(ListNode* headA, ListNode* headB) {
        unordered_set<ListNode*> seen;

        // store all nodes of listA
        while (headA != nullptr) {
            seen.insert(headA);
            headA = headA->next;
        }

        // traverse nodes of listB
        while (headB != nullptr) {
            if (seen.count(headB)) return headB;  // intersection
            headB = headB->next;
        }

        return nullptr;  // no intersection
    }

    ListNode* getIntersectionNodeBetter(ListNode* headA, ListNode* headB) {
        ListNode* ptrA = headA;
        ListNode* ptrB = headB;

        // when ptrA and ptrB traverse both the lists entirely, they travel the
        // same distance, hence they become equal at either:
        // 1: intersection node
        // 2: nullptr (no intersection node)
        while (ptrA != ptrB) {
            ptrA = (ptrA == nullptr) ? headB : ptrA->next;
            ptrB = (ptrB == nullptr) ? headA : ptrB->next;
        }

        // when loop is over, a = b = nullptr (no intersection), or a = b = intersection node
        return ptrA;
    }
};

File: linked_list/150_intersection_of_two_ll.py
# 160. Intersection of Two Linked Lists
from include.linked_list import ListNode
from typing import Optional


class Solution:
    def getIntersectionNode(
        self, headA: ListNode | None, headB: ListNode | None
    ) -> Optional[ListNode]:
        seen: set[ListNode] = set()

        while headA is not None:
            seen.add(headA)
            headA = headA.next

        while headB is not None:
            if headB in seen:
                return headB

            headB = headB.next

        return None

    def getIntersectionNodeBetter(
        self, headA: ListNode | None, headB: ListNode | None
    ) -> Optional[ListNode]:
        ptrA = headA
        ptrB = headB

        while ptrA is not ptrB:
            ptrA = headB if ptrA is None else ptrA.next
            ptrB = headA if ptrB is None else ptrB.next

        return ptrA
